### ISSUE TITLE
NaN values in initial training

### ISSUE THREAD

**Issue by Calvin-Pang** [choosehappy/PatchSorter#28]
Hi, 

I am trying to implement it in some pathology images. However, I did not provide masks or csv files when I first uploaded the images, as I do not have those labels. This brings me NaN values in the 'Train DL' step, as in `train_dl_simclr.py`, the loss function includes a classification part `class_loss=10*criterion_pred(pred_class, gt)` where `gt` will always be -1 if not given in masks or csv files. -1 labels may give the CE loss NaN value and make the model unable to train. Could you please clarify this point? Or is it necessary to have some initial labels before we start the labeling process with PatchSorter?

Thank you so much!

**Comment by choosehappy** [comment]
if you look at the definition of the criterion_pred:

https://github.com/choosehappy/PatchSorter/blob/77456834f76414e7dc4b4e9a49bf7f4e368672f8/patchsorter/approaches/simclr/train_dl_simclr.py#L231


you'll see that the class index of -1 is ignored - thus only those which are labeled are taken into account; if none are labeled, this loss is then 0/ignored. this is very frequently the case when starting a project, so i suspect that your error is elsewhere

can you share your log files?

**Comment by Calvin-Pang** [comment]
Thanks a lot for the clarification! I checked the doc of the CrossEntropyLoss and found it defaults to `reduction='mean'`:

![Image](https://github.com/user-attachments/assets/028dfcad-8ed2-4e94-89aa-c75f0b285520)

Hence, in the initial training stage, it skips all the '-1' labels, and I think it results in zero samples to be divided in the mean operation and leading to `NaN`? 

I also checked the result if I changed it to `reduction='sum'` and it gave me `loss=0`, which makes sense as it will not divide the loss by 0.

**Comment by danielfffan** [comment]
Hi Calvin, thanks for reaching out!

Which version of PyTorch are you using? My project is configured using the Docker environment, which includes PyTorch 1.8.1. This version behaves slightly differently from the latest release (e.g., 2.7.1). We recommend using the provided Docker setup to ensure reproducibility and consistency, especially since it avoids this specific issue.

In PyTorch 2.7.1, as noted in the [official documentation](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html), CrossEntropyLoss expects targets to be class indices. Then the mathematical expression should be the first following one instead the snapshot you put above. If all targets are set to -1 and ignore_index=-1 is used, the entire batch is skipped during loss computation. With reduction='mean', this results in a NaN loss due to division by zero (no valid elements in the batch). 

![Image](https://github.com/user-attachments/assets/cab1a4b6-3b52-4797-a8a8-873e97845f1e)

In contrast, [PyTorch 1.8.1](https://pytorch.org/docs/1.8.1/generated/torch.nn.CrossEntropyLoss.html) handles this case differently. Even if the -1 labels are ignored during loss calculation, it appears the denominator still accounts for the batch size, which results in a loss value of 0.0 rather than NaN.

So, if you're using PyTorch 2.7.1 and seeing NaN values during early training (when no labels are available), this behavior is expected. You can resolve it by one of the following:

- Set reduction='sum' instead of 'mean'

- Add a check to skip loss computation when all targets are -1

- Use the Docker container with PyTorch 1.8.1 (as I do for my project)

Here’s a quick demo using 1.8.1:

```python
>>> import torch.nn as nn
>>> print(torch.__version__)
1.8.1+cu111
>>> torch.manual_seed(0)
<torch._C.Generator object at 0x724b99fcf610>
>>> pred_class = torch.randn(64, 2)
>>> gt = torch.full((pred_class.shape[0],), -1)
>>> criterion = nn.CrossEntropyLoss(ignore_index=-1, weight=None)
>>> criterion(pred_class, gt)
tensor(0.)
```

**Comment by Calvin-Pang** [comment]
Thanks for the clarification! Yes, I did doubt this is due to the difference between the PyTorch versions. Mine is 2.1.0, which may cause this issue. Right now, I use 2nd of your solutions, where I manually add a check to skip the -1 labels. It works well now!
